File: Source/BKBrowser.m
//
//  BKBrowser.m
//  CustomBrowserKit
//
//  Created by ky1vstar on 03/25/2018.
//  Copyright (c) 2018 ky1vstar. All rights reserved.
//

#import "BKBrowser.h"
#import "BKManager.h"
#import "BKBrowserInfo.h"

static NSBundle *frameworkBundle() {
    static dispatch_once_t once;
    static NSBundle *bundle;
    dispatch_once(&once, ^{
        bundle = [NSBundle bundleForClass:[BKManager class]];
        NSLog(@"bundle: %@", bundle);
    });
    return bundle;
}

NSString *BKBrowserGetIdentifier(BKBrowser browser) {
    switch (browser) {
        case BKBrowserInAppSafari:
            return @"inAppSafari";
            
        case BKBrowserSafari:
            return @"safari";
            
        default:
            return [BKBrowserInfo browserInfoForBrowser:browser].identifier;
    }
}

BKBrowser BKBrowserFromIdentifier(NSString *identifier) {
    for (BKBrowser i = 0; i <= BKBrowserSafari; i++)
        if ([BKBrowserGetIdentifier(i) isEqualToString:identifier])
            return i;
    
    return BKBrowserSafari;
}

NSString *BKBrowserGetName(BKBrowser browser) {
    switch (browser) {
        case BKBrowserInAppSafari:
            return @"In-App Safari";
            
        case BKBrowserSafari:
            return @"Safari";
            
        default:
            return [BKBrowserInfo browserInfoForBrowser:browser].name;
    }
}

UIImage *BKBrowserGetThumbnail(BKBrowser browser) {
    return [UIImage imageNamed:BKBrowserGetIdentifier(browser) inBundle:frameworkBundle() compatibleWithTraitCollection:nil];
}

File: Source/BKBrowser.swift
//
//  BKBrowser.swift
//  CustomBrowserKit
//
//  Created by ky1vstar on 03/25/2018.
//  Copyright (c) 2018 ky1vstar. All rights reserved.
//

import Foundation

extension BKBrowser: CustomStringConvertible {

    /// Initialize browser with specified identifier string. It can be used to restore `BKBrowser` from `UserDefaults`.
    /// - note: If identifier is not supported, `.safari` will be returned
    public init(identifier: String) {
        self = __BKBrowserFromIdentifier(identifier)
    }
    
    /// Identifier string for the specified browser. It can be used to save `BKBrowser` to `UserDefaults`.
    public var identifier: String {
        return __BKBrowserGetIdentifier(self)
    }
    
    /// Name of the current browser.
    public var name: String {
        return __BKBrowserGetName(self)
    }
    
    /// Thumbnail image for the current browser.
    /// - note: `nil` if current browser is `.inAppSafari`, image of size 100x100 otherwise
    public var thumbnail: UIImage? {
        return __BKBrowserGetThumbnail(self)
    }
    
    public var description: String {
        return "BKBrowser.\(identifier)"
    }
    
}
